snapshot/kaniko_snapshot_unix_test.go

```diff
@@ -0,0 +1,157 @@
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/sclevine/spec"
+
+	"github.com/buildpacks/lifecycle/snapshot"
+	h "github.com/buildpacks/lifecycle/testhelpers"
+)
+
+func TestKanikoSnapshotter(t *testing.T) {
+	spec.Run(t, "Test Image", testKanikoSnapshotter)
+}
+
+func testKanikoSnapshotter(t *testing.T, when spec.G, it spec.S) {
+	var (
+		snapshotter *snapshot.KanikoSnapshotter
+		tmpDir      string
+	)
+
+	it.Before(func() {
+		// Using the default tmp dir causes kaniko to go haywire for some reason
+		cwd, err := os.Getwd()
+		if err != nil {
+			t.Fatalf("Error: %s\n", err)
+		}
+		tmpDir, err = ioutil.TempDir(filepath.Join(cwd, "..", "tmp"), "kaniko")
+		if err != nil {
+			t.Fatalf("Error: %s\n", err)
+		}
+
+		createTestFile(t, filepath.Join(tmpDir, "cnb", "privatefile"))
+		createTestFile(t, filepath.Join(tmpDir, "layers", "privatefile"))
+		createTestFile(t, filepath.Join(tmpDir, "file-to-change"))
+		createTestFile(t, filepath.Join(tmpDir, "file-not-to-change"))
+		createTestFile(t, filepath.Join(tmpDir, "file-to-delete"))
+		createTestFile(t, filepath.Join(tmpDir, "bin", "file-not-to-change"))
+
+		snapshotter = &snapshot.KanikoSnapshotter{
+			RootDir:                    tmpDir,
+			DetectFilesystemIgnoreList: false,
+		}
+		snapshotter.IgnoredPaths = []string{filepath.Join(snapshotter.RootDir, "dir-with-ignored-files")}
+	})
+
+	it.After(func() {
+		os.RemoveAll(tmpDir)
+	})
+
+	when("files are added and modified", func() {
+		var (
+			snapshotFile string
+		)
+
+		it.Before(func() {
+			h.AssertNil(t, snapshotter.Init())
+
+			os.Remove(filepath.Join(snapshotter.RootDir, "file-to-delete"))
+			createTestFileWithContent(t, filepath.Join(snapshotter.RootDir, "file-to-change"), "hola\n")
+			createTestFile(t, filepath.Join(snapshotter.RootDir, "my-space", "newfile-in-dir"))
+			createTestFile(t, filepath.Join(snapshotter.RootDir, "dir-with-ignored-files", "file-to-ignore"))
+
+			tmpFile, err := ioutil.TempFile("", "snapshot")
+			if err != nil {
+				t.Fatalf("Error: %s\n", err)
+			}
+
+			snapshotFile = tmpFile.Name()
+
+			err = snapshotter.TakeSnapshot(snapshotFile)
+			if err != nil {
+				t.Fatalf("Error: %s\n", err)
+			}
+		})
+
+		it("includes the changed files in the snapshot", func() {
+			data, err := os.Open(snapshotFile)
+			if err != nil {
+				t.Fatalf("Error: %s\n", err)
+			}
+			defer data.Close()
+
+			tr := tar.NewReader(data)
+			for {
+				hdr, err := tr.Next()
+				if err == io.EOF {
+					break // End of archive
+				}
+
+				if err != nil {
+					t.Fatalf("Error: %s\n", err)
+				}
+
+				switch hdr.Name {
+				case "/":
+				case "dir-with-ignored-files/":
+				case "my-space/":
+				case strings.Trim(filepath.Join(snapshotter.RootDir, ".wh.file-to-delete"), "/"):
+					continue
+				case "newfile":
+				case "my-space/newfile-in-dir":
+					assertSnapshotFile(t, tr, "hello\n")
+				case "file-to-change":
+					assertSnapshotFile(t, tr, "hola\n")
+				default:
```

Comment:
This test might be a bit more readable if you had a `case` for `file-to-ignore` and then fall through to any other unexpected files.

Reply:
Hmm I was thinking it would look like 

```
case "layers/file-to-ignore":
  t.Fatalf("Unexpected file: %s\n", hdr.Name)
case "cnb/file-to-ignore":
  t.Fatalf("Unexpected file: %s\n", hdr.Name)
case "platform/file-to-ignore":
  t.Fatalf("Unexpected file: %s\n", hdr.Name)
default:
  t.Fatalf("Unexpected file: %s\n", hdr.Name)
```

I think we would never expect to hit `case "file-to-ignore"` because the test setup never creates that file at the root. But maybe I'm missing something...

Reply:
I cleaned this up a bit more. I realized the case statement also didn't really handle the absence of files and added some assertions around that.